File: generator/src/main/java/ch/bailu/gtk/tag/MethodTag.java
package ch.bailu.gtk.tag;

import java.util.List;

public class MethodTag extends NamedTag {
    private final ParameterTag returnValue =  new ParameterTag(this);

    private final TagList<ParameterTag> parameters = new TagList<>();
    private final TagList<ParameterTag> instanceParameters = new TagList<>();

    private String identifier = "";

    private boolean deprecated = false;
    private boolean throwsError = false;

    public MethodTag(Tag parent) {
        super(parent);
    }


    private class Parameters extends Tag {
        public Parameters() {
            super(MethodTag.this);
        }

        @Override
        public Tag getChild(String name, String prefix) {
            if ("parameter".equals(name)) {
                return parameters.addTag(new ParameterTag(this));
            }
            if ("instance-parameter".equals(name)) {
                return instanceParameters.addTag(new ParameterTag(this));
            }
            return ignore();
        }
    }


    @Override
    public Tag getChild(String name, String prefix) {
        if ("return-value".equals(name)) {
            return returnValue;
        }

        if ("parameters".equals(name)) {
            return new Parameters();
        }

        return new IgnoreTag(this);
    }

    @Override
    public void setAttribute(String name, String value) {
        if ("identifier".equals(name)) {
            identifier = value;
        } else if ("throws".equals(name)) {
            throwsError = "1".equals(value);
        } else if ("deprecated".equals(name)) {
            deprecated = "1".equals(value);
        } else {
            super.setAttribute(name, value);
        }
    }

    public List<ParameterTag> getParameters() {
        return parameters;
    }
    public ParameterTag getReturnValue() {
        return returnValue;
    }

    public String getIdentifier() {
        return identifier;
    }

    public boolean isDeprecated() {
        return deprecated;
    }

    public boolean throwsError() {
        return throwsError;
    }

}

File: generator/src/main/kotlin/ch/bailu/gtk/tag/MethodTag.kt
package ch.bailu.gtk.tag

open class MethodTag(parent: TagWithParent): NamedWithDocTag(parent) {
    private val returnValue = ParameterTag(this)

    private val parameters = TagList<ParameterTag>()
    private val instanceParameters = TagList<ParameterTag>()

    private var identifier = ""

    private var deprecated = false
    private var throwsError = false


    private inner class Parameters : TagWithParent(this@MethodTag) {
        override fun getChild(name: String, prefix: String): TagWithParent {
            if ("parameter" == name) {
                return parameters.addTag(ParameterTag(this))
            }
            if ("instance-parameter" == name) {
                return instanceParameters.addTag(ParameterTag(this))
            }
            return ignore()
        }
    }


    override fun getChild(name: String, prefix: String): TagWithParent {
        if ("return-value" == name) {
            return returnValue
        }
        if ("parameters" == name) {
            return Parameters()
        }
        return super.getChild(name, prefix)
    }

    override fun setAttribute(name: String, value: String) {
        if ("identifier" == name) {
            identifier = value
        } else if ("throws" == name) {
            throwsError = "1" == value
        } else if ("deprecated" == name) {
            deprecated = "1" == value
        } else {
            super.setAttribute(name, value)
        }
    }

    fun getParameters(): List<ParameterTag> {
        return parameters
    }

    fun getReturnValue(): ParameterTag {
        return returnValue
    }

    fun getIdentifier(): String {
        return identifier
    }

    fun isDeprecated(): Boolean {
        return deprecated
    }

    fun throwsError(): Boolean {
        return throwsError
    }
}